### ISSUE TITLE
xpra shadow: on client window maximize, (down)scale the shadowed screen to "perfectly fit" the new window geometry

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#2458]
*Issue migrated from trac ticket # 2458*

**component:** client | **priority:** minor

#### 2019-10-20 13:31:24: stdedos created the issue

___
> Attempt to almost imitate the RDP protocol.
> 
> I would guess that changing the underlaying resolution would be way too complicated, since the monitors are attached to physical monitors.
> 
> When maximizing the client (or going full screen), attempt to scale the image perfectly to size (optionally, give the option to add black bars to avoid weird stretching).
> 
> Note that the scale "you" are doing might be better than the OpenGL scaling (text is unreadable, there are too many artifacts)

**Comment by stdedos** [comment]
I assume this is a priority at the second-tier? (i.e., after https://github.com/Xpra-org/xpra/issues/3948)

Should we use this to track only downscaling (as it was original), or upscaling too?

I assume that after https://github.com/Xpra-org/xpra/issues/3140 `--desktop-scaling` should accept any float value.
The "kicker" would be to do the math, per geometry-change, and handle the N monitors shadowing (if at all possible)

**Comment by stdedos** [comment]
tl;dr: If a server 2560x1440 is rendered at a client 1920x1080 (or smaller, since the xpra client window and the desktop usually take up some of that space too), then xpra shadow server will downscale its output accordingly, to fix the client's viewport exactly (adding black bars as necessary) - without needing specifying `--desktop-scaling` manually

**Comment by stdedos** [comment]
_Even more tl;dr, since "A picture is worth a thousand words"_

![image](https://github.com/user-attachments/assets/1a60f4d2-d56a-4f2b-a122-9a67e92f6c46)

![image](https://github.com/user-attachments/assets/66b4eaed-760c-416d-8280-b70a85551104)

(Rustup)
